src/main/scala/org/broadinstitute/dsde/firecloud/EntityClient.scala

```diff
@@ -111,5 +119,166 @@ class EntityClient (requestContext: RequestContext) extends Actor {
+  //Returns the list of required attributes if all is well.
+  private def withRequiredAttributes(entityType: String, headers: Seq[String])(op: (Map[String, String] => Unit)):Unit = {
+    ModelSchema.getRequiredAttributes(entityType) match {
+      case Failure(regret) => requestContext.complete(HttpResponse(BadRequest, regret.getMessage))
+      case Success(requiredAttributes) => {
+        if( !requiredAttributes.keySet.subsetOf(headers.toSet) ) {
+          requestContext.complete( HttpResponse(BadRequest, "TSV is missing required attributes: " + (requiredAttributes.keySet -- headers).mkString(", ")) )
+        } else {
+          op(requiredAttributes)
+        }
+      }
+    }
+  }
+
+  def batchCallToRawls( workspaceNamespace: String, workspaceName: String, calls: Seq[EntityUpdateDefinition] ) = {
+    log.info("TSV upload request received")
+    val pipeline: HttpRequest => Future[HttpResponse] =
+      addHeader(Cookie(requestContext.request.cookies)) ~> sendReceive
+    val responseFuture: Future[HttpResponse] = pipeline {
+      Post(FireCloudConfig.Workspace.entityPathFromWorkspace(workspaceNamespace, workspaceName)+"/batchUpsert",
+            HttpEntity(MediaTypes.`application/json`,calls.toJson.toString))
+    }
+
+    responseFuture onComplete {
+      case Success(response) =>
+        response.status match {
+          case NoContent =>
+            log.debug("OK response")
+            requestContext.complete(OK)
+          case _ =>
+            // Bubble up all other unmarshallable responses
+            log.warning("Unanticipated response: " + response.status.defaultMessage)
+            requestContext.complete(response)
+        }
+      case Failure(error) =>
+        // Failure accessing service
+        log.error(error, "Service API call failed")
+        requestContext.failWith(
+          new RequestProcessingException(StatusCodes.InternalServerError, error.getMessage))
+    }
+  }
+
+  val upsertAttrOperation = "op" -> AttributeString("AddUpdateAttribute")
+  val removeAttrOperation = "op" -> AttributeString("RemoveAttribute")
+  val addListMemberOperator = "op" -> AttributeString("AddListMember")
+
+  def setAttributesOnEntity(entityType: String, row: Seq[String], colInfo: Seq[(String,Option[String])]) = {
+    val ops = for { (value,(attributeName,refTypeOpt)) <- row.tail zip colInfo if refTypeOpt.isDefined || !value.isEmpty } yield {
+      val nameEntry = "attributeName" -> AttributeString(attributeName)
+      def valEntry( attr: Attribute ) = "addUpdateAttribute" -> attr
+      refTypeOpt match {
+        case Some(refType) => Map(upsertAttrOperation,nameEntry,valEntry(AttributeReference(refType,value)))
+        case None => value match {
+          case "__DELETE__" => Map(removeAttrOperation,nameEntry)
+          case _ => Map(upsertAttrOperation,nameEntry,valEntry(AttributeString(value)))
+        }
+      }
+    }
+    EntityUpdateDefinition(row(0),entityType,ops)
```

Comment:
`headOption` is even better because it won't blow up when empty, and then you can handle that case monadically. Even in a case like this where it appears it's well known that there's a `head` there's still a cognitive load attached to making sure that's been asserted previously (heck, even if it's on the same line).